<div style="background:#FFFFEE; color:#440404; padding:8px; border-radius: 4px; text-align: center; font-weight: 500;">IFN619 - Data Analytics for Strategic Decision Makers</div>

# IFN619 :: UA1 - Assignment 1 - Foundational techniques (35%)

**IMPORTANT:** Refer to the instructions in Canvas module [UA1 - Assignment 1]() *BEFORE* working on this assignment. Ensure that you are familiar with the marking rubric and understand how the grade for this assignment will be awarded according to the criteria.

1. Complete and run the code cell below to display your name and student number
2. Complete all questions in Part A (by week 6) - you can get assistance from your tutor with this part.
3. Ensure that your tutor has verified your understanding of your work in Part A (no later than week 7)
4. Complete a full analysis for Part B. Ensure that you use the techniques and libraries/packages that have been used in class.
5. Check that you have addressed all of the criteria in the assignment rubric
6. Clear all cells and re-run your entire notebook so that cells are sequentially numbered without any errors. **IMPORTANT: Not doing this step risks having your work marked as incomplete!**
7. Submit the clean notebook to Canvas **before** 11:55pm the due date


In [3]:
# Complete the following cell with your details and run to produce your personalised header for this assignment

from IPython.display import HTML

first_name = "Vaishnav"
last_name = "Rai"
student_number = "N11484209"

personal_header = f"<h1>{first_name} {last_name} ({student_number})</h1>"
HTML(personal_header)

---

## Data for both Part A and Part B

This assignment uses data from the Queensland Government [Open Data Portal](https://www.data.qld.gov.au). Both parts will use data on [Queensland Wave Monitoring](https://www.qld.gov.au/environment/coasts-waterways/beach/monitoring). Part B will also use data on [Storm tide monitoring](https://www.qld.gov.au/environment/coasts-waterways/beach/storm) You should familiarise yourself with the information on this site to understand the context for the data.

For this assignment, you will use the [Coastal Data System - Near real time wave data]((https://www.data.qld.gov.au/dataset/coastal-data-system-near-real-time-wave-data)) and for Part B you will add [Coastal Data System – Near real time storm tide data](https://www.data.qld.gov.au/dataset/coastal-data-system-near-real-time-storm-tide-data). Note that this data will change over the time period of the assignment.

---
## Part A

#### QUESTION: 
***What can we learn from the wave height data for South East Queensland, and how might this data be used strategically during a major weather event?***

> *IMPORTANT* For the following task, keep a record of the dates and times where you demonstrated your understanding with your tutor. These should be AFTER you have completed the questions, and BEFORE week 7. Record these below:

**Demo for tutor:** |Tuesday| |01/04/2025| |12:00| (Tutorial 8)

### [Q1] Read the data

The data for this analysis comes from the [Queensland Government's Coastal Data System]((https://www.data.qld.gov.au/dataset/coastal-data-system-near-real-time-wave-data)), which maintains wave monitoring buoys at various locations along the Queensland coast. I accessed this real-time wave data on **March 28, 2025**, capturing a 7-day window of wave measurements. Understanding wave characteristics (height, period, direction) is crucial because they directly influence coastal processes and potential impacts.
When first examining the data, I found it contained 7,663 records with 15 columns including:

* Site location and site number
* DateTime of measurement
* Geographic coordinates (Latitude/Longitude)
* Wave measurements (Hsig - significant wave height, Hmax - maximum wave height)
* Wave period metrics (Tp - peak period, Tz - zero crossing period)
* Direction data (wave direction, current direction)
* Water conditions (SST - sea surface temperature, current speed)

**Justification:** Loading necessary libraries: `pandas` for data manipulation and analysis, `plotly.express` and `plotly.graph_objects` for creating interactive visualizations, and `datetime` for handling date and time operations. Importing `make_subplots` allows for creating more complex figures, such as dual-axis plots used later in Part B.

In [10]:
import pandas as pd
import plotly.express as px
from datetime import datetime
# Import tools for dual-axis plots
from plotly.subplots import make_subplots
import plotly.graph_objects as go

First access date: 2025-03-28

**Justification:** Recording the date when the data was first accessed is crucial for reproducibility, especially when dealing with data sources that are updated over time. This timestamp will help readers understand the specific 7-day window that was analyzed in this notebook.

In [12]:
# [Q1] Read the data
# Make a note of the first access date
first_access_date = "2025-03-28"
print(f"First access date: {first_access_date}")

First access date: 2025-03-28


**Justification:** Initially attempting to load data directly from the source URL. This code is commented out in favour of loading a saved file (next cell) to ensure analysis' consistency and reproducibility, as the live data feed changes. 

---

The wave data is then read into a dataframe using `pd.read_csv`. I discovered the first row had information that caused misalignment of columns. Found code for skipping rows `skiprows=1` from [pandas' read_csv documentation]((https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html))
This was needed to remove/skip past the metadata header that was in the first row, due to which the columns were not being read.

In [14]:
# Open the CSV version of the file directly from the URL into a pandas dataframe
# url = "https://apps.des.qld.gov.au/data-sets/waves/wave-7dayopdata.csv"

In [15]:
# wave_df = pd.read_csv(url, skiprows=1)

**Justification:** Using a local CSV ensures reproducibility, because the live data feed changes over time. Loading the wave data from a previously saved local CSV file (`wave_data_2025-03-28.csv`). This ensures that the analysis uses a consistent, static dataset corresponding to the access date noted earlier (March 28, 2025), making the results reproducible regardless of any changes in the live data source since then.

In [17]:
# Instead of loading from URL, load from our saved file
saved_wave_file = "wave_data_2025-03-28.csv"
wave_df = pd.read_csv(saved_wave_file)
print(f"Using saved wave data from: {saved_wave_file}")

Using saved wave data from: wave_data_2025-03-28.csv


**Justification:** Displaying the first few rows of the loaded wave dataframe using `.head()`. This allows for an initial visual inspection of the data's structure, column names (e.g., 'Hsig', 'Hmax', 'Tp'), and sample values, confirming that the data has been loaded correctly into the pandas DataFrame structure.

In [19]:
# Display the first few rows
print("First 5 rows of the data:")
wave_df.head()

First 5 rows of the data:


id             DateTime       Site SiteNumber     Seconds  Latitude  \
0   1  2025-03-21 00:00:00  Caloundra         54  1742479200 -26.84682   
1   2  2025-03-21 00:30:00  Caloundra         54  1742481000 -26.84671   
2   3  2025-03-21 01:00:00  Caloundra         54  1742482800 -26.84677   
3   4  2025-03-21 01:30:00  Caloundra         54  1742484600 -26.84676   
4   5  2025-03-21 02:00:00  Caloundra         54  1742486400 -26.84658   

   Longitude   Hsig  Hmax    Tp     Tz    SST  Direction  Current Speed  \
0  153.15536  1.226  1.76  9.09  5.195  25.65       92.8          -99.9   
1  153.15536  1.145  2.09  8.33  5.263  25.60       95.6          -99.9   
2  153.15536  1.191  2.00  9.09  5.063  25.55       95.6          -99.9   
3  153.15534  1.103  1.97  9.09  5.333  25.55       85.8          -99.9   
4  153.15540  1.144  1.86  7.69  5.333  25.55       94.2          -99.9   

   Current Direction  
0              -99.9  
1              -99.9  
2              -99.9  
3              -99.9  
4              -99.9

**Justification:** Cleaning the column names of the wave dataframe by removing any leading or trailing whitespace using `.str.strip()`. This ensures consistency in column references and prevents potential errors that can arise from hidden whitespace when selecting or manipulating columns later in the analysis.

In [21]:
# Clean column names by stripping whitespace
wave_df.columns = wave_df.columns.str.strip()

**Justification:** Checking the dimensions (number of rows and columns) using `.shape` and listing all column names using `.columns.tolist()`. This provides a quick overview of the dataset's size and verifies the variables available for analysis are as expected after loading and cleaning.

In [23]:
# Check the shape and columns
print(f"Data shape: {wave_df.shape}")
print(f"Columns: {wave_df.columns.tolist()}")

Data shape: (7663, 15)
Columns: ['id', 'DateTime', 'Site', 'SiteNumber', 'Seconds', 'Latitude', 'Longitude', 'Hsig', 'Hmax', 'Tp', 'Tz', 'SST', 'Direction', 'Current Speed', 'Current Direction']


**Justification:** Examining the data types (`dtypes`) of each column. This is essential to ensure that numerical columns (like 'Hsig', 'Tp') are recognized as numbers (float64/int64) and categorical columns (like 'Site') are appropriate. This step identifies the 'DateTime' column as being of type 'object' (string), indicating it needs conversion to a proper datetime format.

In [25]:
# Check data types to understand the structure
print("\nData types:")
wave_df.dtypes


Data types:


id                     int64
DateTime              object
Site                  object
SiteNumber            object
Seconds                int64
Latitude             float64
Longitude            float64
Hsig                 float64
Hmax                 float64
Tp                   float64
Tz                   float64
SST                  float64
Direction            float64
Current Speed        float64
Current Direction    float64
dtype: object

**Justification:** Converting the 'DateTime' column from its current 'object' type to a pandas datetime format using `pd.to_datetime`. Using `format='mixed'` allows pandas to infer the format, providing robustness if there are minor inconsistencies in the source data's date strings. Displaying `.head()` of the converted column confirms the successful transformation.

---

**Regarding Index Column:** In this analysis, I did not use 'DateTime' or any other column as my DataFrame index because they are not guaranteed to be unique and multiple sites record data at the same time. Certain pandas operations (like some forms of resampling or merging) behave unpredictably when indices are non-unique.

Instead, I kept 'DateTime' as a normal column and used the original 'id' to ensure each row remains uniquely identifiable. This approach simplifies comparisons across different sites and timestamps without forcing a strictly unique time index. If I need to do time-based queries or visualizations, I can still filter or group by the 'DateTime' column directly.

In [27]:
# Convert DateTime to proper datetime format
wave_df['DateTime'] = pd.to_datetime(wave_df['DateTime'], format='mixed')
print("\nDatetime conversion successful. Sample dates:")
wave_df['DateTime'].head()


Datetime conversion successful. Sample dates:


0   2025-03-21 00:00:00
1   2025-03-21 00:30:00
2   2025-03-21 01:00:00
3   2025-03-21 01:30:00
4   2025-03-21 02:00:00
Name: DateTime, dtype: datetime64[ns]

**Justification:** This commented out code is to save the current `wave_df` into a CSV file so that we can save it locally, and call the CSV file in the cells ahead. This ensures reproducibilty and consistency in analysis.

In [29]:
# [Q2] Save the data
# Create a filename with today's date
#filename = f"wave_data_{first_access_date}.csv"
#wave_df_with_index.to_csv(filename)
#print(f"\nData saved to {filename}")

### Read the data from a file

**Justification:** Reading the previously saved wave data CSV back into a new dataframe (`saved_df`). This serves as a check to confirm that the file saving process (like in the previous commented-out cell) worked correctly and the data can be retrieved. For the main analysis continuity, operations will proceed using the primary dataframe (`wave_df`).

In [31]:
# Read the data back from file
saved_df = pd.read_csv("wave_data_2025-03-28.csv")
print("Data successfully read back from file.")

Data successfully read back from file.


### [Q3] Analyse the data


**Justification:** Printing all unique site names present in the dataset provides a complete list of available monitoring locations, which helps in selecting the relevant sites for the South East Queensland focus required by the analysis question. We need to isolate which monitoring stations lie in the South East Queensland. This helps us focus on the wave data most relevant to emergency management in that area.

This initial examination shows us that the dataset contains records from 20 unique monitoring stations along the Queensland coast. 

In [34]:
# [Q3] Analyze the data
# First, let's check the unique sites to identify South East coast sites
print("Available sites:")
for site in sorted(wave_df['Site'].unique()):
    print(f"- {site}")

Available sites:
- Albatross Bay
- Bilinga
- Brisbane Mk4
- Bundaberg
- Cairns Mk4
- Caloundra
- Emu Park
- Gladstone
- Gold Coast Mk4
- Hay Point TriAxys
- Mackay Mk4
- Mooloolaba
- North Moreton Bay
- Palm Beach Mk4
- Poruma West
- Skardon River Outer
- Townsville
- Tweed Heads Mk4
- Tweed Offshore
- Wide Bay


**Justification:** Defining a list (`se_qld_sites`) containing the specific site names relevant to the South East Queensland coast (Gold Coast to Sunshine Coast) as identified for this analysis. This list will be used to filter the main dataframe. I filtered the data to include only the relevant coastal sites. For this analysis, I'll focus specifically on the 9 stations in South East Queensland to address our regional question. I used the [Wave Monitoring Website](https://www.qld.gov.au/environment/coasts-waterways/beach/monitoring/waves-sites) to find the sites on the South East Queensland coast. The website clearly names the wave stations that are in Sunshine Coast, Brisbane, and Gold Coast.

In [61]:
# Define South East Queensland sites (Gold Coast to Sunshine Coast)
se_qld_sites = [
    'Brisbane Mk4', 'Caloundra', 'Gold Coast Mk4', 'Mooloolaba', 'North Moreton Bay' , 
    'Tweed Heads Mk4','Tweed Offshore', 'Palm Beach Mk4', 'Bilinga'
]

**Justification:** Filtering the wave dataframe (`wave_df`) based on the predefined list of relevant sites (`se_qld_sites`). This creates a new dataframe (`se_qld_df`) containing only the wave measurements for the South East Queensland region. Displaying the shape of the filtered data and the count of records per site (`.value_counts()`) confirms the filtering was successful and provides an overview of the data available for each targeted location. We can see that all of our sites have a similar number of records, with discrepancies of 1-5 data points missing at few sites like North Moreton Bay. Palm Beach Mk4 and Bilinga amongst others have 379 records. This near-complete coverage is beneficial for emergency managers, as data gaps might hide peak wave events.

In [64]:
# Filter data for matched SE QLD sites
se_qld_df = wave_df[wave_df['Site'].isin(se_qld_sites)]

# Show the filtered data size
print(f"\nFiltered data shape: {se_qld_df.shape}")
print(f"Number of records by site:")
print(se_qld_df['Site'].value_counts())


Filtered data shape: (3400, 15)
Number of records by site:
Site
Tweed Heads Mk4      379
Palm Beach Mk4       379
Gold Coast Mk4       379
Bilinga              379
Mooloolaba           378
Brisbane Mk4         378
Tweed Offshore       378
Caloundra            376
North Moreton Bay    374
Name: count, dtype: int64


**Justification:** Grouping the filtered South East Queensland wave data (`se_qld_df`) by 'Site' and calculating key aggregate statistics for each site over the entire 7-day period. The aggregates include mean, maximum, and minimum Significant Wave Height ('Hsig'), mean and maximum Maximum Wave Height ('Hmax'), mean Peak Period ('Tp'), and mean Direction. This process condenses thousands of time-series records into a manageable summary table, allowing for direct comparison of typical conditions, peak conditions, wave periods, and directions across the different SE QLD sites. `.reset_index()` converts the grouped output back into a standard dataframe. Flattening the hierarchical column index (created by the multi-level aggregation) makes column names simpler and easier to use. Sorting the resulting dataframe by mean Significant Wave Height (`Hsig mean`) arranges the sites from highest to lowest average wave exposure that helps us with interpretation.

---

***Note**: GenAI (Claude) was used to assist in generating the code for flattening the hierarchical column index after aggregation and for sorting by the mean Hsig.*

**Choice and Value of Aggregations:**

To effectively summarize the wave climate during the observation period (March 21-28, 2025) and extract meaningful insights, the following aggregations were calculated by site.

---

*   **Significant Wave Height (Mean, Max, Min):** Provides a measure of the average wave energy (Mean), the peak intensity observed (Max), and baseline conditions (Min). Crucial for understanding both typical exposure and extreme event characteristics.

---

*   **Maximum Wave Height (Mean, Max):** Offers insight into the typical size of the largest waves (Mean Hmax) and the single highest wave recorded (Max Hmax), important for safety thresholds and structural considerations.

---

*   **Peak Period (Mean):** Indicates the dominant wave period, helping to infer the nature of the waves (e.g., shorter period wind waves vs. longer period swell). Longer period waves generally carry more energy.

---

*   **Wave Direction (Mean):** Shows the predominant direction waves approached from, critical for understanding coastal alignment impacts and potential erosion patterns.

---

**Strategic Value:** These aggregations offer a multi-dimensional perspective on the wave environment. They enable comparison of sites based on typical and extreme wave exposure, provide context for assessing coastal risk, help inform resource allocation during storm events by identifying higher-impact zones, and contribute to public safety messaging by characterizing wave conditions.  By identifying the site with the highest average Hsig and the largest max Hmax, we can see which coastal areas are likely most impacted during a severe storm. We will discuss more during the insights section

---

In [68]:
# Group by site and calculate mean wave heights and other statistics
grouped_df = se_qld_df.groupby('Site').agg({
    'Hsig': ['mean', 'max', 'min'],  # Significant wave height
    'Hmax': ['mean', 'max'],        # Maximum wave height
    'Tp': 'mean',                   # Peak period
    'Direction': 'mean'             # Wave direction
}).reset_index()
# Flatten the column hierarchy
grouped_df.columns = [' '.join(col).strip() if col[1] else col[0] for col in grouped_df.columns.values]

# Sort by mean significant wave height (descending)
grouped_df = grouped_df.sort_values('Hsig mean', ascending=False)

print("\nWave statistics by site in South East Queensland:")
grouped_df


Wave statistics by site in South East Queensland:


Site  Hsig mean  Hsig max  Hsig min  Hmax mean  Hmax max  \
4         Mooloolaba   1.413862     2.462     0.740   2.445450      4.88   
1       Brisbane Mk4   1.336058     2.230     0.760   2.267011      4.19   
8     Tweed Offshore   1.246243     1.980     0.720   2.107831      3.67   
5  North Moreton Bay   1.144957     2.035     0.623   1.972032      3.70   
7    Tweed Heads Mk4   1.117757     1.790     0.630   1.886781      3.24   
2          Caloundra   1.108457     1.913     0.591   1.889229      3.34   
3     Gold Coast Mk4   1.104433     1.750     0.590   1.871979      3.29   
6     Palm Beach Mk4   1.030739     1.740     0.570   1.758391      3.16   
0            Bilinga   0.927441     1.600     0.480   1.567731      3.12   

    Tp mean  Direction mean  
4  7.140370       89.240212  
1  7.063201       80.632302  
8  7.810159      107.193254  
5  7.075829       75.054813  
7  7.755673       94.068285  
2  7.096410       83.257979  
3  8.008654       89.093140  
6  6.806464       79.144063  
0  6.692902       70.497942

This reveals both typical conditions and extremes for each location over the observed week (March 21-28, 2025). Key patterns relevant to coastal management include:

1.  **Exposure:** A clear gradient exists, with Mooloolaba (mean Significant Wave Height 1.41m) and Brisbane Mk4 (1.34m) showing the highest average wave energy exposure in this dataset. Bilinga (0.93m) consistently experienced the most sheltered conditions.

---

2.  **Peak Event Intensity:** Mooloolaba recorded the highest significant wave height (max 2.46m) which likely occurred around March 26 afternoon (we will confirm in a time-series plot down below). This might coincide with a local storm event. It also received highest maximum wave height (max 4.88m).

---


3.  **Wave Period:** Mean peak periods mostly range from 6.69s to 8.01s, suggesting a mix of conditions potentially including local wind waves and some swell.

---


4.  **Predominant Direction:** Mean wave directions generally fall within the East to North-East sector (approx. 70° to 94° for most sites, Tweed Offshore 107°).

---
5.  **Synchronized Patterns (See Time Series Plot):** The timing of wave height changes appears broadly synchronized across sites, suggesting the influence of a regional weather system affecting the coastline.


---

**Justification:** This commented-out code block demonstrates how the *aggregated* wave statistics dataframe (`grouped_df`) could be saved to a CSV file. This preserves the summarized results derived from the analysis for reporting or later use.

In [72]:
# Save the grouped data to a CSV with the date
# Not Necessary but for personal use
# grouped_filename = f"wave_data_grouped_{first_access_date}.csv"
# grouped_df.to_csv(grouped_filename, index=False)
# print(f"Grouped data saved to {grouped_filename}")